proof/refine/ARM/Ipc_R.thy

```diff
@@ -4141,240 +4281,198 @@ lemma tcbSchedEnqueue_valid_tcb_state'[wp]:
   "tcbSchedEnqueue t \<lbrace>valid_tcb_state' ts\<rbrace>"
   by (wpsimp simp: tcbSchedEnqueue_def)
 
-crunches rescheduleRequired, scheduleTCB
-  for valid_tcb_state'[wp]: "valid_tcb_state' ts"
-  (wp: crunch_wps)
-
-crunches setThreadState
-  for sc_at'_n[wp]: "sc_at'_n n p"
-  and sc_ko_at'[wp]: "\<lambda>s. P (ko_at' (sc :: sched_context) p s)"
-  (wp: crunch_wps simp: crunch_simps)
-
-definition replyPush' where
```

Comment:
I have been cleaning things up with this PR (and mine to remove overlaps), and noticed that there are a fair bit of unnecessary changes introduced in this PR. I think we want to revert and cleanup those.

Firstly, I don’t see why we need to remove `replyPush’`. This is Ryan’s work that led to a nice wp rule for `valid_objs’` and others. Consequently, in the below, these wp rules have added preconditions that we shouldn’t need. Please re-establish Ryan’s results for these wp rules, and for that, I believe resurrecting `replyPush’` is the shortest path.

Reply:
I thought that the spec change in this PR is semantically equivalent, but I can now see that there is not just the `bindScReply` part but also the part where asserts are removed. I understand that this does mean that we might now need `reply_at’ replyPtr` in the precondition. But I think it should be just that and we shouldn't need `tcb_at’` for `caller` and `callee`. The `list_refs_of_replies’` lemma below also doesn’t need any of the added preconditions, I believe.
And, at this point, I am not entirely sure how much `replyPush’` can help in these. It is certainly possible to update it with the spec change for `replyPush`, but I now think that it is most likely that the issue of redundant preconditions can be sorted out without it.

Reply:
Ok, cool I'll check it out and maybe we can chat more later today.

Reply:
I think I agree with Miki's second comment. I admit that I didn't understand the `replyPush'` proof very well, but the changes in this PR made other work much easier. In particular, I was finding it extremely hard to prove `replyPush_valid_replies'` before this PR was merged, and it was nice and straightforward after.
If we could strengthen the lemmas by weakening the preconditions then that would clearly be nice, but I'd be very sad to see this reverted, and without reverting I think resurrecting `replyPush'` might be tricky.